File: macos/API/AudioFormat.swift
import Foundation
import AVFoundation

/// AudioFormat - Represents an audio format configuration
///
/// This structure encapsulates all parameters needed to describe an audio format,
/// including sample rate, channel configuration, bit depth, and data layout.
public struct AudioFormat: Equatable, Hashable, Codable {
    /// Sample rate in Hz
    public let sampleRate: Double
    
    /// Number of channels
    public let channelCount: UInt32
    
    /// Bit depth (16, 24, 32)
    public let bitDepth: UInt32
    
    /// Whether samples are interleaved
    public let isInterleaved: Bool
    
    /// Whether format uses floating point
    public let isFloat: Bool
    
    /// Bytes per frame
    public var bytesPerFrame: UInt32 {
        return (bitDepth / 8) * (isInterleaved ? channelCount : 1)
    }
    
    /// Bytes per packet
    public var bytesPerPacket: UInt32 {
        return bytesPerFrame * (isInterleaved ? 1 : channelCount)
    }
    
    /// Common format type
    public var commonFormat: AVAudioCommonFormat {
        switch (bitDepth, isFloat) {
        case (16, false):
            return .pcmFormatInt16
        case (32, true):
            return .pcmFormatFloat32
        case (32, false):
            return .pcmFormatInt32
        case (24, false):
            return isInterleaved ? .pcmFormatInt24 : .pcmFormatInt32
        default:
            return .pcmFormatFloat32
        }
    }
    
    /// Default initializer
    public init(sampleRate: Double, channelCount: UInt32, bitDepth: UInt32, isInterleaved: Bool, isFloat: Bool = false) {
        self.sampleRate = sampleRate
        self.channelCount = channelCount
        self.bitDepth = bitDepth
        self.isInterleaved = isInterleaved
        self.isFloat = isFloat || bitDepth == 32 // Default to float for 32-bit
    }
    
    /// Create from AVAudioFormat
    public init?(from avFormat: AVAudioFormat) {
        self.sampleRate = avFormat.sampleRate
        self.channelCount = avFormat.channelCount
        self.isInterleaved = avFormat.isInterleaved
        
        switch avFormat.commonFormat {
        case .pcmFormatInt16:
            self.bitDepth = 16
            self.isFloat = false
        case .pcmFormatInt32:
            self.bitDepth = 32
            self.isFloat = false
        case .pcmFormatFloat32:
            self.bitDepth = 32
            self.isFloat = true
        case .pcmFormatFloat64:
            self.bitDepth = 64
            self.isFloat = true
        default:
            return nil
        }
    }
    
    /// Convert to AVAudioFormat
    public func toAVAudioFormat() -> AVAudioFormat? {
        return AVAudioFormat(
            commonFormat: commonFormat,
            sampleRate: sampleRate,
            channels: channelCount,
            interleaved: isInterleaved
        )
    }
    
    /// Check if format is compatible with another format
    public func isCompatible(with other: AudioFormat) -> Bool {
        return self.sampleRate == other.sampleRate &&
               self.channelCount == other.channelCount &&
               self.bitDepth == other.bitDepth &&
               self.isInterleaved == other.isInterleaved &&
               self.isFloat == other.isFloat
    }
    
    /// Check if format requires conversion to another format
    public func requiresConversion(to other: AudioFormat) -> Bool {
        return !isCompatible(with: other)
    }
    
    /// Human-readable description
    public var description: String {
        let formatType = isFloat ? "Float" : "Int"
        let layout = isInterleaved ? "Interleaved" : "Non-interleaved"
        return "\(Int(sampleRate))Hz, \(channelCount)ch, \(bitDepth)-bit \(formatType), \(layout)"
    }
    
    // MARK: - Common Formats
    
    /// Default capture format (48kHz, 2ch, 32-bit float, non-interleaved)
    public static let defaultFormat = AudioFormat(
        sampleRate: 48000.0,
        channelCount: 2,
        bitDepth: 32,
        isInterleaved: false,
        isFloat: true
    )
    
    /// CD quality format (44.1kHz, 2ch, 16-bit int, interleaved)
    public static let cdQuality = AudioFormat(
        sampleRate: 44100.0,
        channelCount: 2,
        bitDepth: 16,
        isInterleaved: true,
        isFloat: false
    )
    
    /// Standard WAV format (48kHz, 2ch, 16-bit int, interleaved)
    public static let standardWAV = AudioFormat(
        sampleRate: 48000.0,
        channelCount: 2,
        bitDepth: 16,
        isInterleaved: true,
        isFloat: false
    )
    
    /// High quality format (96kHz, 2ch, 24-bit int, interleaved)
    public static let highQuality = AudioFormat(
        sampleRate: 96000.0,
        channelCount: 2,
        bitDepth: 24,
        isInterleaved: true,
        isFloat: false
    )
}

/// AudioFormatNegotiator - Handles format negotiation and conversion
public class AudioFormatNegotiator {
    
    /// Find best common format between source and destination
    public static func negotiate(source: AudioFormat, destination: AudioFormat, preferences: FormatPreferences = .default) -> AudioFormat {
        // If formats match, no negotiation needed
        if source.isCompatible(with: destination) {
            return source
        }
        
        // Apply preferences
        switch preferences.priority {
        case .quality:
            // Prefer higher sample rate and bit depth
            return AudioFormat(
                sampleRate: max(source.sampleRate, destination.sampleRate),
                channelCount: max(source.channelCount, destination.channelCount),
                bitDepth: max(source.bitDepth, destination.bitDepth),
                isInterleaved: destination.isInterleaved,
                isFloat: source.isFloat || destination.isFloat
            )
            
        case .compatibility:
            // Prefer destination format for maximum compatibility
            return destination
            
        case .performance:
            // Prefer source format to minimize conversion
            return source
            
        case .balanced:
            // Find middle ground
            return AudioFormat(
                sampleRate: destination.sampleRate, // Use destination sample rate
                channelCount: min(source.channelCount, destination.channelCount),
                bitDepth: destination.bitDepth,
                isInterleaved: destination.isInterleaved,
                isFloat: destination.isFloat
            )
        }
    }
    
    /// Check if direct conversion is possible
    public static func canConvert(from source: AudioFormat, to destination: AudioFormat) -> Bool {
        // Check for supported conversions
        if source.sampleRate != destination.sampleRate {
            // Sample rate conversion is supported
            return true
        }
        
        if source.channelCount != destination.channelCount {
            // Channel count mismatch - check if downmix/upmix is possible
            if source.channelCount > 2 && destination.channelCount == 2 {
                return true // Can downmix to stereo
            }
            if source.channelCount == 1 && destination.channelCount == 2 {
                return true // Can upmix mono to stereo
            }
        }
        
        // Format conversion is generally possible
        return true
    }
    
    /// Get conversion complexity score (0.0 = simple, 1.0 = complex)
    public static func conversionComplexity(from source: AudioFormat, to destination: AudioFormat) -> Float {
        var complexity: Float = 0.0
        
        // Sample rate conversion
        if source.sampleRate != destination.sampleRate {
            complexity += 0.3
        }
        
        // Channel conversion
        if source.channelCount != destination.channelCount {
            complexity += 0.2
        }
        
        // Bit depth conversion
        if source.bitDepth != destination.bitDepth {
            complexity += 0.2
        }
        
        // Float/Int conversion
        if source.isFloat != destination.isFloat {
            complexity += 0.2
        }
        
        // Interleaving conversion
        if source.isInterleaved != destination.isInterleaved {
            complexity += 0.1
        }
        
        return min(complexity, 1.0)
    }
}

/// Format negotiation preferences
public struct FormatPreferences {
    /// Negotiation priority
    public enum Priority {
        case quality        // Prefer highest quality
        case compatibility  // Prefer most compatible format
        case performance    // Prefer least conversion
        case balanced       // Balance all factors
    }
    
    /// Priority for negotiation
    public let priority: Priority
    
    /// Maximum sample rate to consider
    public let maxSampleRate: Double?
    
    /// Maximum bit depth to consider
    public let maxBitDepth: UInt32?
    
    /// Prefer interleaved formats
    public let preferInterleaved: Bool
    
    /// Prefer floating point formats
    public let preferFloat: Bool
    
    /// Default preferences
    public static let `default` = FormatPreferences(
        priority: .balanced,
        maxSampleRate: nil,
        maxBitDepth: nil,
        preferInterleaved: true,
        preferFloat: false
    )
    
    /// High quality preferences
    public static let highQuality = FormatPreferences(
        priority: .quality,
        maxSampleRate: 192000.0,
        maxBitDepth: 32,
        preferInterleaved: false,
        preferFloat: true
    )
    
    /// Performance preferences
    public static let performance = FormatPreferences(
        priority: .performance,
        maxSampleRate: 48000.0,
        maxBitDepth: 16,
        preferInterleaved: true,
        preferFloat: false
    )
}

/// Audio buffer wrapper for format-aware operations
public struct AudioBuffer {
    /// The underlying PCM buffer
    public let pcmBuffer: AVAudioPCMBuffer
    
    /// The audio format
    public let format: AudioFormat
    
    /// Timestamp of buffer capture
    public let timestamp: Date
    
    /// Duration in seconds
    public var duration: TimeInterval {
        return Double(pcmBuffer.frameLength) / format.sampleRate
    }
    
    /// Create from PCM buffer
    public init(pcmBuffer: AVAudioPCMBuffer, timestamp: Date = Date()) {
        self.pcmBuffer = pcmBuffer
        self.format = AudioFormat(from: pcmBuffer.format) ?? AudioFormat.defaultFormat
        self.timestamp = timestamp
    }
    
    /// Create with format and frame capacity
    public init?(format: AudioFormat, frameCapacity: AVAudioFrameCount) {
        guard let avFormat = format.toAVAudioFormat(),
              let buffer = AVAudioPCMBuffer(pcmFormat: avFormat, frameCapacity: frameCapacity) else {
            return nil
        }
        self.pcmBuffer = buffer
        self.format = format
        self.timestamp = Date()
    }
}

File: windows/API/AudioFormat.py
"""
AudioFormat - Represents an audio format configuration

This module encapsulates all parameters needed to describe an audio format,
including sample rate, channel configuration, bit depth, and data layout.
"""

from dataclasses import dataclass
from datetime import datetime
from enum import Enum, auto
from typing import Optional, Tuple
import numpy as np


class AudioCommonFormat(Enum):
    """Common audio format types"""
    PCM_FORMAT_INT16 = auto()
    PCM_FORMAT_INT32 = auto()
    PCM_FORMAT_FLOAT32 = auto()
    PCM_FORMAT_FLOAT64 = auto()


@dataclass(frozen=True)
class AudioFormat:
    """
    Represents an audio format configuration.
    
    Attributes:
        sample_rate: Sample rate in Hz
        channel_count: Number of channels
        bit_depth: Bit depth (16, 24, 32, 64)
        is_interleaved: Whether samples are interleaved
        is_float: Whether format uses floating point
    """
    sample_rate: float
    channel_count: int
    bit_depth: int
    is_interleaved: bool
    is_float: bool = False
    
    def __post_init__(self):
        """Validate and adjust format parameters"""
        if self.bit_depth == 32 and not self.is_float:
            # Default to float for 32-bit
            object.__setattr__(self, 'is_float', True)
    
    @property
    def bytes_per_frame(self) -> int:
        """Calculate bytes per frame"""
        return (self.bit_depth // 8) * (self.channel_count if self.is_interleaved else 1)
    
    @property
    def bytes_per_packet(self) -> int:
        """Calculate bytes per packet"""
        return self.bytes_per_frame * (1 if self.is_interleaved else self.channel_count)
    
    @property
    def common_format(self) -> AudioCommonFormat:
        """Get common format type"""
        if self.bit_depth == 16 and not self.is_float:
            return AudioCommonFormat.PCM_FORMAT_INT16
        elif self.bit_depth == 32 and self.is_float:
            return AudioCommonFormat.PCM_FORMAT_FLOAT32
        elif self.bit_depth == 32 and not self.is_float:
            return AudioCommonFormat.PCM_FORMAT_INT32
        elif self.bit_depth == 64 and self.is_float:
            return AudioCommonFormat.PCM_FORMAT_FLOAT64
        elif self.bit_depth == 24 and not self.is_float:
            # 24-bit is typically converted to 32-bit
            return AudioCommonFormat.PCM_FORMAT_INT32
        else:
            return AudioCommonFormat.PCM_FORMAT_FLOAT32
    
    @property
    def numpy_dtype(self) -> np.dtype:
        """Get NumPy dtype for this format"""
        if self.bit_depth == 16 and not self.is_float:
            return np.dtype('int16')
        elif self.bit_depth == 32 and self.is_float:
            return np.dtype('float32')
        elif self.bit_depth == 32 and not self.is_float:
            return np.dtype('int32')
        elif self.bit_depth == 64 and self.is_float:
            return np.dtype('float64')
        else:
            return np.dtype('float32')
    
    def is_compatible(self, other: 'AudioFormat') -> bool:
        """Check if format is compatible with another format"""
        return (self.sample_rate == other.sample_rate and
                self.channel_count == other.channel_count and
                self.bit_depth == other.bit_depth and
                self.is_interleaved == other.is_interleaved and
                self.is_float == other.is_float)
    
    def requires_conversion(self, other: 'AudioFormat') -> bool:
        """Check if format requires conversion to another format"""
        return not self.is_compatible(other)
    
    @property
    def description(self) -> str:
        """Human-readable description"""
        format_type = "Float" if self.is_float else "Int"
        layout = "Interleaved" if self.is_interleaved else "Non-interleaved"
        return f"{int(self.sample_rate)}Hz, {self.channel_count}ch, {self.bit_depth}-bit {format_type}, {layout}"
    
    # Common format presets
    @classmethod
    def default_format(cls) -> 'AudioFormat':
        """Default capture format (48kHz, 2ch, 32-bit float, non-interleaved)"""
        return cls(
            sample_rate=48000.0,
            channel_count=2,
            bit_depth=32,
            is_interleaved=False,
            is_float=True
        )
    
    @classmethod
    def cd_quality(cls) -> 'AudioFormat':
        """CD quality format (44.1kHz, 2ch, 16-bit int, interleaved)"""
        return cls(
            sample_rate=44100.0,
            channel_count=2,
            bit_depth=16,
            is_interleaved=True,
            is_float=False
        )
    
    @classmethod
    def standard_wav(cls) -> 'AudioFormat':
        """Standard WAV format (48kHz, 2ch, 16-bit int, interleaved)"""
        return cls(
            sample_rate=48000.0,
            channel_count=2,
            bit_depth=16,
            is_interleaved=True,
            is_float=False
        )
    
    @classmethod
    def high_quality(cls) -> 'AudioFormat':
        """High quality format (96kHz, 2ch, 24-bit int, interleaved)"""
        return cls(
            sample_rate=96000.0,
            channel_count=2,
            bit_depth=24,
            is_interleaved=True,
            is_float=False
        )


class FormatPriority(Enum):
    """Format negotiation priority"""
    QUALITY = auto()        # Prefer highest quality
    COMPATIBILITY = auto()  # Prefer most compatible format
    PERFORMANCE = auto()    # Prefer least conversion
    BALANCED = auto()       # Balance all factors


@dataclass
class FormatPreferences:
    """Format negotiation preferences"""
    priority: FormatPriority
    max_sample_rate: Optional[float] = None
    max_bit_depth: Optional[int] = None
    prefer_interleaved: bool = True
    prefer_float: bool = False
    
    @classmethod
    def default(cls) -> 'FormatPreferences':
        """Default preferences"""
        return cls(
            priority=FormatPriority.BALANCED,
            max_sample_rate=None,
            max_bit_depth=None,
            prefer_interleaved=True,
            prefer_float=False
        )
    
    @classmethod
    def high_quality(cls) -> 'FormatPreferences':
        """High quality preferences"""
        return cls(
            priority=FormatPriority.QUALITY,
            max_sample_rate=192000.0,
            max_bit_depth=32,
            prefer_interleaved=False,
            prefer_float=True
        )
    
    @classmethod
    def performance(cls) -> 'FormatPreferences':
        """Performance preferences"""
        return cls(
            priority=FormatPriority.PERFORMANCE,
            max_sample_rate=48000.0,
            max_bit_depth=16,
            prefer_interleaved=True,
            prefer_float=False
        )


class AudioFormatNegotiator:
    """Handles format negotiation and conversion"""
    
    @staticmethod
    def negotiate(
        source: AudioFormat,
        destination: AudioFormat,
        preferences: FormatPreferences = None
    ) -> AudioFormat:
        """Find best common format between source and destination"""
        if preferences is None:
            preferences = FormatPreferences.default()
        
        # If formats match, no negotiation needed
        if source.is_compatible(destination):
            return source
        
        # Apply preferences
        if preferences.priority == FormatPriority.QUALITY:
            # Prefer higher sample rate and bit depth
            return AudioFormat(
                sample_rate=max(source.sample_rate, destination.sample_rate),
                channel_count=max(source.channel_count, destination.channel_count),
                bit_depth=max(source.bit_depth, destination.bit_depth),
                is_interleaved=destination.is_interleaved,
                is_float=source.is_float or destination.is_float
            )
        
        elif preferences.priority == FormatPriority.COMPATIBILITY:
            # Prefer destination format for maximum compatibility
            return destination
        
        elif preferences.priority == FormatPriority.PERFORMANCE:
            # Prefer source format to minimize conversion
            return source
        
        else:  # BALANCED
            # Find middle ground
            return AudioFormat(
                sample_rate=destination.sample_rate,  # Use destination sample rate
                channel_count=min(source.channel_count, destination.channel_count),
                bit_depth=destination.bit_depth,
                is_interleaved=destination.is_interleaved,
                is_float=destination.is_float
            )
    
    @staticmethod
    def can_convert(source: AudioFormat, destination: AudioFormat) -> bool:
        """Check if direct conversion is possible"""
        # Sample rate conversion is supported
        if source.sample_rate != destination.sample_rate:
            return True
        
        # Channel count mismatch - check if downmix/upmix is possible
        if source.channel_count != destination.channel_count:
            if source.channel_count > 2 and destination.channel_count == 2:
                return True  # Can downmix to stereo
            if source.channel_count == 1 and destination.channel_count == 2:
                return True  # Can upmix mono to stereo
        
        # Format conversion is generally possible
        return True
    
    @staticmethod
    def conversion_complexity(source: AudioFormat, destination: AudioFormat) -> float:
        """Get conversion complexity score (0.0 = simple, 1.0 = complex)"""
        complexity = 0.0
        
        # Sample rate conversion
        if source.sample_rate != destination.sample_rate:
            complexity += 0.3
        
        # Channel conversion
        if source.channel_count != destination.channel_count:
            complexity += 0.2
        
        # Bit depth conversion
        if source.bit_depth != destination.bit_depth:
            complexity += 0.2
        
        # Float/Int conversion
        if source.is_float != destination.is_float:
            complexity += 0.2
        
        # Interleaving conversion
        if source.is_interleaved != destination.is_interleaved:
            complexity += 0.1
        
        return min(complexity, 1.0)


class AudioBuffer:
    """Audio buffer wrapper for format-aware operations"""
    
    def __init__(
        self,
        data: np.ndarray,
        format: AudioFormat,
        timestamp: Optional[datetime] = None
    ):
        """
        Initialize audio buffer.
        
        Args:
            data: NumPy array containing audio data
            format: Audio format specification
            timestamp: Timestamp of buffer capture
        """
        self.data = data
        self.format = format
        self.timestamp = timestamp or datetime.now()
        
        # Validate data shape
        expected_shape = self._calculate_expected_shape()
        if self.data.shape != expected_shape and len(expected_shape) > 0:
            raise ValueError(
                f"Data shape {self.data.shape} doesn't match format. "
                f"Expected {expected_shape}"
            )
    
    def _calculate_expected_shape(self) -> Tuple[int, ...]:
        """Calculate expected data shape based on format"""
        if len(self.data.shape) == 1:
            # 1D array for mono or interleaved
            return self.data.shape
        
        if self.format.is_interleaved:
            # Interleaved: (frames, channels)
            if len(self.data.shape) == 2:
                return self.data.shape
        else:
            # Non-interleaved: (channels, frames)
            if len(self.data.shape) == 2:
                return self.data.shape
        
        return tuple()
    
    @property
    def frame_count(self) -> int:
        """Get number of audio frames"""
        if self.format.is_interleaved:
            return self.data.shape[0] if len(self.data.shape) > 0 else 0
        else:
            return self.data.shape[1] if len(self.data.shape) > 1 else len(self.data)
    
    @property
    def duration(self) -> float:
        """Get duration in seconds"""
        return self.frame_count / self.format.sample_rate
    
    def to_interleaved(self) -> 'AudioBuffer':
        """Convert to interleaved format"""
        if self.format.is_interleaved:
            return self
        
        # Transpose data from (channels, frames) to (frames, channels)
        interleaved_data = self.data.T
        interleaved_format = AudioFormat(
            sample_rate=self.format.sample_rate,
            channel_count=self.format.channel_count,
            bit_depth=self.format.bit_depth,
            is_interleaved=True,
            is_float=self.format.is_float
        )
        
        return AudioBuffer(interleaved_data, interleaved_format, self.timestamp)
    
    def to_non_interleaved(self) -> 'AudioBuffer':
        """Convert to non-interleaved format"""
        if not self.format.is_interleaved:
            return self
        
        # Transpose data from (frames, channels) to (channels, frames)
        non_interleaved_data = self.data.T
        non_interleaved_format = AudioFormat(
            sample_rate=self.format.sample_rate,
            channel_count=self.format.channel_count,
            bit_depth=self.format.bit_depth,
            is_interleaved=False,
            is_float=self.format.is_float
        )
        
        return AudioBuffer(non_interleaved_data, non_interleaved_format, self.timestamp)